### PR TITLE
Debugger: Reduces RAM useage for searches (Act 1)

### PR DESCRIPTION
**Pull request by Daniel-McCarthy**
### Description of Changes
Reduces memory useage during memory searching process.
- Minimizes passing by copy, uses references to avoid causing copies
- Deletes FutureWatcher ptr when no longer needed to avoid leaking memory (thanks Fobes)
- Updates the search results vector in-place so that no extra copies are needed.
- Makes use of std::move to transfer resources instead of keeping two copies.
- No longer keeps a copy of prior search results on main thread in addition to the active search. The stored results are transferred to the search worker and then transferred back.
- Updates the storage of prior results to use a vector instead of QMap, making indexing for results loading much quicker and less intensive.

More improvements to be made, but making a first that should make a decent impact.

Also fixes an issue where a crash occurred when using searches that require prior results with a New Search by giving a message and refusing the search, letting the user that the search comparison only works with `Filter Search`. Thanks @kamfretoz for catching this!

### Rationale behind Changes
Storing prior results with debugger searches to be able to check prior state based comparisons (such as increased/decreased/changed) caused a quite noticeable increase in ram usage that needs to be resolved.

### Suggested Testing Steps
Validating searches are still working as expected and comparing peak ram usage from main branch to this branch.

### CODE REVIEW

**Comment by kamfretoz** [comment]
Peak memory usage still looks to be rather high.

Screenshot_20240309_201719

**Comment by Daniel-McCarthy** [comment]
> Peak memory usage still looks to be rather high.
It is still high, but it's a lot lower than it was on the prior version. This was meant to be more of a quick improvement to reduce impact in the mean time while I'm working on a solution that will address it fully.

**Review by F0bes — APPROVED** [review]
The fix wasn't as great as I thought it'd be. I spent a good hour messing with Visual Studio's heap profiler (that breaks with a large amount of memory usage) and didn't come to a silver bullet of a solution.
There is still a copy somewhere, let's say the results take 3GB, the total memory usage for a search will peak at 6GB.
The QMap clear appears to take O(n) time. When added on to the time to search and copy, is a very long time.

Thankfully, unlike before, clearing the search results results in the previous map data being destroyed instead of leaking. 

I definitely think using a vector type will be better here. Discarding the results will be much, much faster.
This will address the fact that the ui freezes when you change the search type (due to it clearing the `m_searchResultsMap`).

When you swap over to using something other than a vector, I might open up the profiler again and take a look.

Otherwise, LGTM.

**Comment by Daniel-McCarthy** [comment]
> There is still a copy somewhere, let's say the results take 3GB, the total memory usage for a search will peak at 6GB.

This will be fixed by switching to a vector, it's at the loading of results into the UI. 

> I definitely think using a vector type will be better here. Discarding the results will be much, much faster. This will address the fact that the ui freezes when you change the search type (due to it clearing the `m_searchResultsMap`).

I will do that, I originally didn't swap to it because I had some rougher changes that didn't make as much impact, however when I worked on this branch, I added some approaches I didn't try with the vector one. I'll swap it to using a vector and try to get it cleaned up. It should really just go into this PR anyhow. 👍

**Review by kamfretoz — APPROVED** [review]

**Comment by Daniel-McCarthy** [comment]
@F0bes I have updated it to use the `std::vector` again. I have applied this as a new commit (that I will merge later) so that the new changes can be compared with the original PR implementation.

This should have a positive impact for the loading of results as well as the mentioned clearing times. Any feedback would certainly be appreciated.

**Review by kamfretoz — APPROVED** [review]
The crash has been fixed.

image

File: pcsx2-qt/Debugger/MemorySearchWidget.cpp [hunk 1]
```diff
@@ -544,38 +556,48 @@ void MemorySearchWidget::onSearchButtonClicked()
...
+	connect(workerWatcher, &QFutureWatcher<std::vector<SearchResult>>::finished, onSearchFinished);
+	m_searchResults.clear();
 	m_ui.resultsCountLabel->setText(tr("Searching..."));
 	m_ui.resultsCountLabel->setVisible(true);
 }
 
 void MemorySearchWidget::onSearchResultsListScroll(u32 value)
 {
-	const bool hasResultsToLoad = static_cast<qsizetype>(m_ui.listSearchResults->count()) < m_searchResultsMap.size();
+	const bool hasResultsToLoad = static_cast<qsizetype>(m_ui.listSearchResults->count()) < m_searchResults.size();
```
Comment (F0bes) [inline pcsx2-qt/Debugger/MemorySearchWidget.cpp:600]: Change the static cast to `size_t`

**Review by F0bes — APPROVED** [review]
I tested and didn't even need to bring the memory profiler out, it works great.
~1GB of memory usage for ~26 million results. As soon as the results are cleared the memory is freed. Good job.

I haven't tested it thoroughly, but so far it looks good to me.

**Review by F0bes — APPROVED** [review]

**Review by Mrlinkwii — APPROVED** [review]

**Pull request merged**